File: tests/verify-freelancer-tax.js
/**
 * Manual verification of FreelancerTaxCalc accuracy
 * Test case: ₪30,000/month gross revenue, no expenses,
 *            5% pension, 4.5% keren, 2.25 credit points, Osek Patur
 *
 * Run: node tests/verify-freelancer-tax.js
 */

// ─── 2026 constants (from tax-data.js) ───────────────────────────
const TAX_BRACKETS = [
    { limit: 7010,     rate: 0.10 },
    { limit: 10060,    rate: 0.14 },
    { limit: 19000,    rate: 0.20 },
    { limit: 25100,    rate: 0.31 },
    { limit: 46690,    rate: 0.35 },
    { limit: 60130,    rate: 0.47 },
    { limit: Infinity, rate: 0.50 }
];

const NI_THRESHOLD    = 7522;
const NI_CEILING      = 48281;
const SELF_NI_LOWER   = 0.0287;
const SELF_NI_UPPER   = 0.1283;
const SELF_HL_LOWER   = 0.031;
const SELF_HL_UPPER   = 0.05;
const CREDIT_PT_VALUE = 242;

const SE = {
    NI_DEDUCTION_RATE:             0.52,
    PENSION_RECOGNIZED_INCOME:     293400 / 12,   // 24,450
    PENSION_DEDUCTION_RATE:        0.11,
    PENSION_CREDIT_RATE:           0.055,
    KEREN_MAX_DEPOSIT_RATE:        0.07,
    KEREN_DEDUCTIBLE_RATE:         0.045,
    KEREN_INCOME_CEILING:          293400 / 12,
};

// ─── Helpers ─────────────────────────────────────────────────────
const fmt = (n) => '₪' + n.toLocaleString('he-IL', { maximumFractionDigits: 2 });
const pct = (n) => (n * 100).toFixed(2) + '%';

function calcIncomeTax(monthlyTaxable) {
    let tax = 0, prev = 0;
    for (const b of TAX_BRACKETS) {
        if (monthlyTaxable <= prev) break;
        tax += (Math.min(monthlyTaxable, b.limit) - prev) * b.rate;
        prev = b.limit;
    }
    return tax;
}

function calcNI(gross) {
    const capped = Math.min(gross, NI_CEILING);
    if (capped <= NI_THRESHOLD) return capped * SELF_NI_LOWER;
    return NI_THRESHOLD * SELF_NI_LOWER + (capped - NI_THRESHOLD) * SELF_NI_UPPER;
}

function calcHealth(gross) {
    const capped = Math.min(gross, NI_CEILING);
    if (capped <= NI_THRESHOLD) return capped * SELF_HL_LOWER;
    return NI_THRESHOLD * SELF_HL_LOWER + (capped - NI_THRESHOLD) * SELF_HL_UPPER;
}

// ─── Test Inputs ─────────────────────────────────────────────────
const MONTHLY_REVENUE  = 30000;
const ANNUAL_REVENUE   = MONTHLY_REVENUE * 12;      // 360,000
const ANNUAL_EXPENSES  = 0;
const PENSION_PCT      = 5;                         // %
const KEREN_PCT        = 4.5;                       // %
const CREDIT_POINTS    = 2.25;

console.log('═══════════════════════════════════════════════════════');
console.log('  MANUAL TAX VERIFICATION — ₪30,000/month freelancer');
console.log('  Pension 5% | Keren 4.5% | 2.25 credit pts | Osek Patur');
console.log('═══════════════════════════════════════════════════════\n');

// ─── Step 1: Gross → Taxable before NI ───────────────────────────
const annualTaxableBeforeNI = ANNUAL_REVENUE - ANNUAL_EXPENSES;
const monthlyTaxableBeforeNI = annualTaxableBeforeNI / 12;
console.log('STEP 1 — Revenue after expenses');
console.log(`  Annual taxable before NI: ${fmt(annualTaxableBeforeNI)}  (${fmt(monthlyTaxableBeforeNI)}/mo)\n`);

// ─── Step 2: NI & Health (on monthly gross after expenses) ───────
const monthlyNI     = calcNI(monthlyTaxableBeforeNI);
const monthlyHealth = calcHealth(monthlyTaxableBeforeNI);
const annualNI      = monthlyNI * 12;
const annualHealth  = monthlyHealth * 12;

console.log('STEP 2 — Bituach Leumi & Health');
const niLower  = NI_THRESHOLD * SELF_NI_LOWER;
const niUpper  = (Math.min(MONTHLY_REVENUE, NI_CEILING) - NI_THRESHOLD) * SELF_NI_UPPER;
console.log(`  NI lower band:  ${fmt(NI_THRESHOLD)} × ${pct(SELF_NI_LOWER)} = ${fmt(niLower)}`);
console.log(`  NI upper band:  ${fmt(MONTHLY_REVENUE - NI_THRESHOLD)} × ${pct(SELF_NI_UPPER)} = ${fmt(niUpper)}`);
console.log(`  Monthly NI:     ${fmt(monthlyNI)}`);
console.log(`  Annual NI:      ${fmt(annualNI)}`);
const hlLower = NI_THRESHOLD * SELF_HL_LOWER;
const hlUpper = (Math.min(MONTHLY_REVENUE, NI_CEILING) - NI_THRESHOLD) * SELF_HL_UPPER;
console.log(`  Health lower:   ${fmt(NI_THRESHOLD)} × ${pct(SELF_HL_LOWER)} = ${fmt(hlLower)}`);
console.log(`  Health upper:   ${fmt(MONTHLY_REVENUE - NI_THRESHOLD)} × ${pct(SELF_HL_UPPER)} = ${fmt(hlUpper)}`);
console.log(`  Monthly health: ${fmt(monthlyHealth)}`);
console.log(`  Annual health:  ${fmt(annualHealth)}\n`);

// ─── Step 3: 52% NI deduction ────────────────────────────────────
const niDeduction = annualNI * SE.NI_DEDUCTION_RATE;
console.log('STEP 3 — 52% NI deduction from taxable income');
console.log(`  ${fmt(annualNI)} × 52% = ${fmt(niDeduction)}\n`);

// ─── Step 4: Pension deduction ───────────────────────────────────
const recognizedMonthly        = Math.min(monthlyTaxableBeforeNI, SE.PENSION_RECOGNIZED_INCOME);
const pensionDeductPct         = Math.min(PENSION_PCT / 100, SE.PENSION_DEDUCTION_RATE);
const pensionMonthlyDeduction  = recognizedMonthly * pensionDeductPct;
const annualPensionDeduction   = pensionMonthlyDeduction * 12;
const pensionMonthlyContrib    = recognizedMonthly * (PENSION_PCT / 100);
const annualPensionContrib     = pensionMonthlyContrib * 12;
// Credit: 5.5% of contribution at 35%
const pensionCreditAmount      = recognizedMonthly * SE.PENSION_CREDIT_RATE * 0.35 * 12;

console.log('STEP 4 — Pension deduction & credit');
console.log(`  Recognized monthly income (capped at ${fmt(SE.PENSION_RECOGNIZED_INCOME)}): ${fmt(recognizedMonthly)}`);
console.log(`  Monthly pension contribution (5%): ${fmt(pensionMonthlyContrib)}`);
console.log(`  Deductible (min(5%,11%)=${pct(pensionDeductPct)}): ${fmt(pensionMonthlyDeduction)}/mo = ${fmt(annualPensionDeduction)}/yr`);
console.log(`  Pension tax credit (5.5% × 35%): ${fmt(pensionCreditAmount)}/yr\n`);

// ─── Step 5: Keren Hishtalmut ────────────────────────────────────
const kerenCapped           = Math.min(monthlyTaxableBeforeNI, SE.KEREN_INCOME_CEILING);
const kerenMonthlyContrib   = kerenCapped * Math.min(KEREN_PCT / 100, SE.KEREN_MAX_DEPOSIT_RATE);
const kerenMonthlyDeductible = kerenCapped * Math.min(KEREN_PCT / 100, SE.KEREN_DEDUCTIBLE_RATE);
const annualKerenContrib    = kerenMonthlyContrib * 12;
const annualKerenDeduction  = kerenMonthlyDeductible * 12;

console.log('STEP 5 — Keren Hishtalmut (4.5%)');
console.log(`  Income capped at ceiling: ${fmt(kerenCapped)}`);
console.log(`  Monthly contribution (4.5%): ${fmt(kerenMonthlyContrib)}`);
console.log(`  Monthly deductible (min(4.5%,4.5%)): ${fmt(kerenMonthlyDeductible)}`);
console.log(`  Annual Keren deduction: ${fmt(annualKerenDeduction)}\n`);

// ─── Step 6: Adjusted taxable income ────────────────────────────
const annualAdjustedTaxable  = Math.max(0, annualTaxableBeforeNI - niDeduction - annualPensionDeduction - annualKerenDeduction);
const monthlyAdjustedTaxable = annualAdjustedTaxable / 12;

console.log('STEP 6 — Adjusted taxable income');
console.log(`  ${fmt(annualTaxableBeforeNI)} − ${fmt(niDeduction)} (NI) − ${fmt(annualPensionDeduction)} (pension) − ${fmt(annualKerenDeduction)} (keren)`);
console.log(`  Annual adjusted: ${fmt(annualAdjustedTaxable)}`);
console.log(`  Monthly adjusted: ${fmt(monthlyAdjustedTaxable)}\n`);

// ─── Step 7: Income tax on adjusted income ───────────────────────
const monthlyIncomeTaxBefore = calcIncomeTax(monthlyAdjustedTaxable);
const annualIncomeTaxBefore  = monthlyIncomeTaxBefore * 12;
const annualCreditAmount     = CREDIT_POINTS * CREDIT_PT_VALUE * 12;
const annualIncomeTax        = Math.max(0, annualIncomeTaxBefore - annualCreditAmount - pensionCreditAmount);

console.log('STEP 7 — Income tax calculation');
console.log(`  Monthly taxable: ${fmt(monthlyAdjustedTaxable)}`);
let prev = 0;
for (const b of TAX_BRACKETS) {
    if (monthlyAdjustedTaxable <= prev) break;
    const taxable = Math.min(monthlyAdjustedTaxable, b.limit) - prev;
    console.log(`  Bracket up to ${fmt(b.limit)}: ${fmt(taxable)} × ${pct(b.rate)} = ${fmt(taxable * b.rate)}`);
    prev = b.limit;
}
console.log(`  Monthly income tax (before credits): ${fmt(monthlyIncomeTaxBefore)}`);
console.log(`  Annual income tax (before credits): ${fmt(annualIncomeTaxBefore)}`);
console.log(`  Credit points deduction: ${CREDIT_POINTS} × ${CREDIT_PT_VALUE} × 12 = ${fmt(annualCreditAmount)}`);
console.log(`  Pension credit: ${fmt(pensionCreditAmount)}`);
console.log(`  Annual income tax (after credits): ${fmt(annualIncomeTax)}\n`);

// ─── Step 8: Totals ──────────────────────────────────────────────
const totalMandatoryTax = annualIncomeTax + annualNI + annualHealth;
const totalDeductions   = totalMandatoryTax + annualPensionContrib + annualKerenContrib;
const annualNet         = ANNUAL_REVENUE - totalDeductions;
const effectiveRate     = (totalMandatoryTax / ANNUAL_REVENUE) * 100;

console.log('═══════════════════════════════════════════════════════');
console.log('  ANNUAL SUMMARY');
console.log('═══════════════════════════════════════════════════════');
console.log(`  Gross revenue:      ${fmt(ANNUAL_REVENUE)}`);
console.log(`  Income tax:         ${fmt(annualIncomeTax)}`);
console.log(`  Bituach Leumi:      ${fmt(annualNI)}`);
console.log(`  Health insurance:   ${fmt(annualHealth)}`);
console.log(`  Pension (5%):       ${fmt(annualPensionContrib)}`);
console.log(`  Keren Hishtalmut:   ${fmt(annualKerenContrib)}`);
console.log(`  ─────────────────────────────────────────`);
console.log(`  Total deductions:   ${fmt(totalDeductions)}`);
console.log(`  NET annual:         ${fmt(annualNet)}`);
console.log(`  NET monthly avg:    ${fmt(annualNet / 12)}`);
console.log(`  Effective tax rate: ${effectiveRate.toFixed(2)}%`);
console.log(`  Marginal rate:      35% (bracket ₪25,100-₪46,690)`);

// ─── Cross-check: quick sanity tests ────────────────────────────
console.log('\n═══════════════════════════════════════════════════════');
console.log('  SANITY CHECKS');
console.log('═══════════════════════════════════════════════════════');
const checks = [
    ['Net < Gross',                annualNet < ANNUAL_REVENUE],
    ['Net > 0',                    annualNet > 0],
    ['Income tax > 0',             annualIncomeTax > 0],
    ['NI > 0',                     annualNI > 0],
    ['Effective rate 20–45%',      effectiveRate > 20 && effectiveRate < 45],
    ['NI deduction < annual NI',   niDeduction < annualNI],
    ['Adjusted taxable < gross',   annualAdjustedTaxable < annualTaxableBeforeNI],
    ['Pension deduction ≤ 11%',    annualPensionDeduction <= annualPensionContrib * 1.001],
    ['Tax + NI + health = total',  Math.abs(annualIncomeTax + annualNI + annualHealth - totalMandatoryTax) < 0.01],
];

let allPassed = true;
for (const [label, result] of checks) {
    console.log(`  ${result ? '✅' : '❌'} ${label}`);
    if (!result) allPassed = false;
}

// ─── Additional test: lower income (₪8,000/month, below NI threshold) ───
console.log('\n═══════════════════════════════════════════════════════');
console.log('  EDGE CASE — ₪8,000/month (straddles NI threshold)');
console.log('═══════════════════════════════════════════════════════');
const m2 = 8000;
const ni2 = calcNI(m2);
const hl2 = calcHealth(m2);
// Manual: lower: 7522×0.0287=215.88; upper: 478×0.1283=61.33
const ni2_expected = NI_THRESHOLD * SELF_NI_LOWER + (m2 - NI_THRESHOLD) * SELF_NI_UPPER;
console.log(`  Monthly NI on ₪8,000:`);
console.log(`    Lower: ${fmt(NI_THRESHOLD)} × ${pct(SELF_NI_LOWER)} = ${fmt(NI_THRESHOLD * SELF_NI_LOWER)}`);
console.log(`    Upper: ${fmt(m2 - NI_THRESHOLD)} × ${pct(SELF_NI_UPPER)} = ${fmt((m2 - NI_THRESHOLD) * SELF_NI_UPPER)}`);
console.log(`    Total: ${fmt(ni2_expected)}  →  calc(): ${fmt(ni2)}`);
console.log(`  ${Math.abs(ni2 - ni2_expected) < 0.01 ? '✅' : '❌'} NI matches expected`);

// ─── Additional test: income below NI threshold (₪5,000/month) ──
console.log('\n  EDGE CASE — ₪5,000/month (below NI threshold)');
const m3 = 5000;
const ni3 = calcNI(m3);
const ni3_expected = m3 * SELF_NI_LOWER;  // only lower rate applies
console.log(`  Monthly NI: ${fmt(m3)} × ${pct(SELF_NI_LOWER)} = ${fmt(ni3_expected)}  →  calc(): ${fmt(ni3)}`);
console.log(`  ${Math.abs(ni3 - ni3_expected) < 0.01 ? '✅' : '❌'} NI matches expected`);

// ─── Additional test: pension cap at 11% ────────────────────────
console.log('\n  EDGE CASE — Pension at 15% (over 11% deduction cap)');
const m4_monthly = 20000;
const m4_recognized = Math.min(m4_monthly, SE.PENSION_RECOGNIZED_INCOME);
const m4_pension_pct = 15;
const m4_deductPct = Math.min(m4_pension_pct / 100, SE.PENSION_DEDUCTION_RATE);
const m4_contribMonthly = m4_recognized * (m4_pension_pct / 100);
const m4_deductMonthly = m4_recognized * m4_deductPct;
console.log(`  Pension 15% of ${fmt(m4_recognized)}: ${fmt(m4_contribMonthly)}/mo`);
console.log(`  Deductible capped at 11%: ${fmt(m4_deductMonthly)}/mo`);
console.log(`  ${m4_deductPct === 0.11 ? '✅' : '❌'} Deduction correctly capped at 11%`);

console.log('\n' + (allPassed ? '✅ ALL CHECKS PASSED' : '❌ SOME CHECKS FAILED'));
console.log('═══════════════════════════════════════════════════════\n');

File: tests/verify-freelancer-tax.py
"""
Manual verification of FreelancerTaxCalc accuracy.
Replicates the exact same logic as js/freelancer-tax.js using 2026 constants.

Run: python3 tests/verify-freelancer-tax.py
"""

# ─── 2026 Constants (mirrored from tax-data.js) ─────────────────
TAX_BRACKETS = [
    (7010,     0.10),
    (10060,    0.14),
    (19000,    0.20),
    (25100,    0.31),
    (46690,    0.35),
    (60130,    0.47),
    (float('inf'), 0.50),
]
NI_THRESHOLD    = 7522
NI_CEILING      = 48281
SELF_NI_LOWER   = 0.0287
SELF_NI_UPPER   = 0.1283
SELF_HL_LOWER   = 0.031
SELF_HL_UPPER   = 0.05
CREDIT_PT_VALUE = 242

SE_NI_DEDUCTION_RATE       = 0.52
SE_PENSION_RECOGNIZED      = 293400 / 12   # 24,450
SE_PENSION_DEDUCT_RATE     = 0.11
SE_PENSION_CREDIT_RATE     = 0.055
SE_KEREN_MAX_RATE          = 0.07
SE_KEREN_DEDUCTIBLE_RATE   = 0.045
SE_KEREN_INCOME_CEILING    = 293400 / 12

# ─── Core helpers (exact replica of JS functions) ─────────────────
def calc_income_tax(monthly_taxable):
    tax, prev = 0.0, 0.0
    for limit, rate in TAX_BRACKETS:
        if monthly_taxable <= prev:
            break
        tax += (min(monthly_taxable, limit) - prev) * rate
        prev = limit
    return tax

def calc_ni(monthly_gross):
    capped = min(monthly_gross, NI_CEILING)
    if capped <= NI_THRESHOLD:
        return capped * SELF_NI_LOWER
    return NI_THRESHOLD * SELF_NI_LOWER + (capped - NI_THRESHOLD) * SELF_NI_UPPER

def calc_health(monthly_gross):
    capped = min(monthly_gross, NI_CEILING)
    if capped <= NI_THRESHOLD:
        return capped * SELF_HL_LOWER
    return NI_THRESHOLD * SELF_HL_LOWER + (capped - NI_THRESHOLD) * SELF_HL_UPPER

def calc_annual_tax(annual_revenue, annual_expenses, pension_pct, keren_pct, credit_points):
    """Exact replica of FreelancerTaxCalc.calculateAnnualTax()"""
    # Step 1
    annual_taxable_before_ni  = max(0, annual_revenue - annual_expenses)
    monthly_taxable_before_ni = annual_taxable_before_ni / 12

    # Step 2: NI & Health
    monthly_ni     = calc_ni(monthly_taxable_before_ni)
    monthly_health = calc_health(monthly_taxable_before_ni)
    annual_ni      = monthly_ni * 12
    annual_health  = monthly_health * 12

    # Step 3: 52% NI deduction
    ni_deduction = annual_ni * SE_NI_DEDUCTION_RATE

    # Step 4: Pension
    # Credit: 35% of min(actual contribution, 5.5% of recognized). Deduction: up to 11%.
    recognized_monthly      = min(monthly_taxable_before_ni, SE_PENSION_RECOGNIZED)
    pension_deduct_pct      = min(pension_pct / 100, SE_PENSION_DEDUCT_RATE)
    pension_monthly_deduct  = recognized_monthly * pension_deduct_pct
    annual_pension_deduct   = pension_monthly_deduct * 12
    pension_monthly_contrib = recognized_monthly * (pension_pct / 100)
    annual_pension_contrib  = pension_monthly_contrib * 12
    pension_credit_amount   = min(pension_monthly_contrib,
                                  recognized_monthly * SE_PENSION_CREDIT_RATE) * 12 * 0.35

    # Step 5: Keren
    keren_capped            = min(monthly_taxable_before_ni, SE_KEREN_INCOME_CEILING)
    keren_monthly_contrib   = keren_capped * min(keren_pct / 100, SE_KEREN_MAX_RATE)
    keren_monthly_deduct    = keren_capped * min(keren_pct / 100, SE_KEREN_DEDUCTIBLE_RATE)
    annual_keren_contrib    = keren_monthly_contrib * 12
    annual_keren_deduct     = keren_monthly_deduct * 12

    # Step 6: Adjusted taxable
    annual_adjusted_taxable  = max(0, annual_taxable_before_ni - ni_deduction
                                      - annual_pension_deduct - annual_keren_deduct)
    monthly_adjusted_taxable = annual_adjusted_taxable / 12

    # Step 7: Income tax
    monthly_tax_before     = calc_income_tax(monthly_adjusted_taxable)
    annual_tax_before      = monthly_tax_before * 12
    annual_credit_amount   = credit_points * CREDIT_PT_VALUE * 12
    annual_income_tax      = max(0, annual_tax_before - annual_credit_amount - pension_credit_amount)

    total_mandatory_tax = annual_income_tax + annual_ni + annual_health
    annual_net          = annual_revenue - annual_expenses - total_mandatory_tax \
                          - annual_pension_contrib - annual_keren_contrib
    effective_rate      = (total_mandatory_tax / annual_revenue * 100) if annual_revenue else 0

    return dict(
        annual_revenue=annual_revenue,
        annual_expenses=annual_expenses,
        annual_taxable_before_ni=annual_taxable_before_ni,
        monthly_taxable_before_ni=monthly_taxable_before_ni,
        monthly_ni=monthly_ni, annual_ni=annual_ni,
        monthly_health=monthly_health, annual_health=annual_health,
        ni_deduction=ni_deduction,
        recognized_monthly=recognized_monthly,
        pension_monthly_contrib=pension_monthly_contrib,
        annual_pension_contrib=annual_pension_contrib,
        annual_pension_deduct=annual_pension_deduct,
        pension_credit_amount=pension_credit_amount,
        annual_keren_contrib=annual_keren_contrib,
        annual_keren_deduct=annual_keren_deduct,
        monthly_adjusted_taxable=monthly_adjusted_taxable,
        annual_adjusted_taxable=annual_adjusted_taxable,
        annual_tax_before=annual_tax_before,
        annual_credit_amount=annual_credit_amount,
        pension_credit=pension_credit_amount,
        annual_income_tax=annual_income_tax,
        total_mandatory_tax=total_mandatory_tax,
        annual_net=annual_net,
        effective_rate=effective_rate,
    )

def fmt(n): return f"₪{n:,.2f}"
def pct(n): return f"{n*100:.2f}%"
def sep(): print("═" * 60)

def run_case(title, monthly_revenue, annual_expenses, pension_pct, keren_pct, credit_points):
    annual_revenue = monthly_revenue * 12
    r = calc_annual_tax(annual_revenue, annual_expenses, pension_pct, keren_pct, credit_points)
    sep()
    print(f"  {title}")
    sep()
    print(f"\n  INPUTS")
    print(f"    Monthly revenue  : {fmt(monthly_revenue)}")
    print(f"    Annual revenue   : {fmt(annual_revenue)}")
    print(f"    Annual expenses  : {fmt(annual_expenses)}")
    print(f"    Pension          : {pension_pct}%")
    print(f"    Keren            : {keren_pct}%")
    print(f"    Credit points    : {credit_points}")

    print(f"\n  STEP-BY-STEP BREAKDOWN")
    print(f"    [1] Taxable before NI  : {fmt(r['annual_taxable_before_ni'])}  ({fmt(r['monthly_taxable_before_ni'])}/mo)")
    print(f"    [2] Monthly NI         : {fmt(r['monthly_ni'])}  →  annual {fmt(r['annual_ni'])}")
    print(f"        Monthly health     : {fmt(r['monthly_health'])}  →  annual {fmt(r['annual_health'])}")
    print(f"    [3] NI deduction (52%) : {fmt(r['ni_deduction'])}")
    print(f"    [4] Pension deduction  : {fmt(r['annual_pension_deduct'])}/yr  (credit: {fmt(r['pension_credit'])})")
    print(f"    [5] Keren deduction    : {fmt(r['annual_keren_deduct'])}/yr")
    print(f"    [6] Adjusted taxable   : {fmt(r['annual_adjusted_taxable'])}  ({fmt(r['monthly_adjusted_taxable'])}/mo)")
    print(f"    [7] Tax before credits : {fmt(r['annual_tax_before'])}")
    print(f"        Credit pts amt     : {fmt(r['annual_credit_amount'])}")
    print(f"        Pension credit     : {fmt(r['pension_credit'])}")
    print(f"        Income tax         : {fmt(r['annual_income_tax'])}")

    print(f"\n  ANNUAL RESULTS")
    print(f"    Income tax         : {fmt(r['annual_income_tax'])}")
    print(f"    Bituach Leumi      : {fmt(r['annual_ni'])}")
    print(f"    Health insurance   : {fmt(r['annual_health'])}")
    print(f"    Total mandatory    : {fmt(r['total_mandatory_tax'])}")
    print(f"    Pension contrib    : {fmt(r['annual_pension_contrib'])}")
    print(f"    Keren contrib      : {fmt(r['annual_keren_contrib'])}")
    print(f"    ─────────────────────────────────────")
    print(f"    NET annual         : {fmt(r['annual_net'])}")
    print(f"    NET monthly avg    : {fmt(r['annual_net']/12)}")
    print(f"    Effective tax rate : {r['effective_rate']:.2f}%")

    # Sanity checks
    checks = [
        ("Net < Gross",                  r['annual_net'] < annual_revenue),
        ("Net > 0",                      r['annual_net'] > 0),
        ("Income tax > 0",               r['annual_income_tax'] > 0),
        ("NI > 0",                       r['annual_ni'] > 0),
        ("Effective rate reasonable",    5 < r['effective_rate'] < 60),
        ("NI deduction < annual NI",     r['ni_deduction'] < r['annual_ni']),
        ("Adjusted taxable < gross",     r['annual_adjusted_taxable'] < r['annual_taxable_before_ni']),
        ("Pension deduction ≤ contrib",  r['annual_pension_deduct'] <= r['annual_pension_contrib'] + 0.01),
        ("Components sum correctly",     abs(r['annual_income_tax'] + r['annual_ni'] + r['annual_health']
                                             - r['total_mandatory_tax']) < 0.01),
    ]
    print(f"\n  SANITY CHECKS")
    all_pass = True
    for label, result in checks:
        icon = "✅" if result else "❌"
        print(f"    {icon} {label}")
        if not result:
            all_pass = False
    return all_pass, r

# ─── Run test cases ──────────────────────────────────────────────
all_passed = True
results = {}

ok, r1 = run_case("TEST 1 — ₪30,000/mo, no expenses, pension 5%, keren 4.5%",
                  monthly_revenue=30000, annual_expenses=0,
                  pension_pct=5, keren_pct=4.5, credit_points=2.25)
all_passed &= ok
results['t1'] = r1

ok, r2 = run_case("TEST 2 — ₪15,000/mo, ₪3,000/mo expenses, no pension/keren",
                  monthly_revenue=15000, annual_expenses=3000*12,
                  pension_pct=0, keren_pct=0, credit_points=2.25)
all_passed &= ok
results['t2'] = r2

ok, r3 = run_case("TEST 3 — ₪50,000/mo, max pension 11%, max keren 7%",
                  monthly_revenue=50000, annual_expenses=0,
                  pension_pct=11, keren_pct=7, credit_points=2.25)
all_passed &= ok
results['t3'] = r3

ok, r4 = run_case("TEST 4 — ₪8,000/mo (near NI threshold), no deductions",
                  monthly_revenue=8000, annual_expenses=0,
                  pension_pct=0, keren_pct=0, credit_points=2.25)
all_passed &= ok
results['t4'] = r4

# ─── Edge case: pension capped at 11% ──────────────────────────
sep()
print("  EDGE CASE — Pension > 11% deduction cap")
sep()
r_cap = calc_annual_tax(20000*12, 0, pension_pct=15, keren_pct=0, credit_points=2.25)
r_uncap = calc_annual_tax(20000*12, 0, pension_pct=11, keren_pct=0, credit_points=2.25)
print(f"  Pension at 15%: deduction = {fmt(r_cap['annual_pension_deduct'])}")
print(f"  Pension at 11%: deduction = {fmt(r_uncap['annual_pension_deduct'])}")
cap_ok = abs(r_cap['annual_pension_deduct'] - r_uncap['annual_pension_deduct']) < 0.01
print(f"  {'✅' if cap_ok else '❌'} Deduction identical at 15% and 11% (capped correctly)")
all_passed &= cap_ok

# ─── Edge case: NI ceiling ─────────────────────────────────────
sep()
print("  EDGE CASE — NI ceiling (₪48,281)")
sep()
ni_at_ceiling  = calc_ni(NI_CEILING)
ni_above_ceil  = calc_ni(NI_CEILING + 10000)
print(f"  NI at ceiling  (₪{NI_CEILING:,}):      {fmt(ni_at_ceiling)}")
print(f"  NI above ceiling (₪{NI_CEILING+10000:,}): {fmt(ni_above_ceil)}")
ceil_ok = abs(ni_at_ceiling - ni_above_ceil) < 0.01
print(f"  {'✅' if ceil_ok else '❌'} NI identical at and above ceiling (cap works)")
all_passed &= ceil_ok

# ─── Edge case: credit pts wipe tax at low income ─────────────
sep()
print("  EDGE CASE — Very low income (₪3,000/mo), should have near-zero tax")
sep()
r_low = calc_annual_tax(3000*12, 0, 0, 0, credit_points=2.25)
print(f"  Income tax: {fmt(r_low['annual_income_tax'])}  (expected ≈ 0 due to credits)")
low_ok = r_low['annual_income_tax'] == 0
print(f"  {'✅' if low_ok else '❌'} Income tax correctly floored at 0")
all_passed &= low_ok

# ─── Cross-reference: Test 1 vs published benchmark ───────────
sep()
print("  CROSS-REFERENCE — Test 1 vs expected ranges")
sep()
# For ₪30K/month freelancer with 5% pension + 4.5% keren:
# Accountant estimate range: income tax ~₪58K-65K, NI ~₪37K, effective ~30-33%
r = results['t1']
# With 5% pension + 4.5% keren deductions applied, income tax for ₪30K/mo is ~₪48-54K
# NI for ₪30K/mo should be ~₪35K-40K, effective rate ~27-32%
print(f"  Income tax     : {fmt(r['annual_income_tax'])}  (expected range ₪45,000–₪58,000)")
print(f"  Bituach Leumi  : {fmt(r['annual_ni'])}  (expected range ₪35,000–₪40,000)")
print(f"  Effective rate : {r['effective_rate']:.1f}%  (expected range 27–34%)")
ref_ok = (45000 <= r['annual_income_tax'] <= 58000 and
          35000 <= r['annual_ni'] <= 40000 and
          27 <= r['effective_rate'] <= 34)
print(f"  {'✅' if ref_ok else '❌'} Results within expected accountant range")
all_passed &= ref_ok

sep()
print(f"\n  {'✅ ALL CHECKS PASSED' if all_passed else '❌ SOME CHECKS FAILED'}")
sep()
